src/Transient/Internals.hs

```diff
@@ -601,264 +600,260 @@ delData x=  modify $ \st -> st{mfData= M.delete (typeOf x ) (mfData st)}
-                    | otherwise= drop (ev:processed) th evts
-
-hangThread parent child = when(not $ freeTh parent) $ do
-   let headpths= children parent
-   atomically $ do
-       ths <- readTVar headpths
-       writeTVar headpths $  child:ths   -- !!>  "thread added: "++ show (threadId child)
-
--- | kill  all the child threads associated with the continuation context
+forkFinally1 action andThen =
+  mask $ \restore -> forkIO $ try (restore action) >>= andThen
+
+-- | Search for the given ThreadId in the continuation context hierarchy and remove -- it from the children list of an ancestor if possible.
+free :: ThreadId -> EventF -> IO (Maybe EventF)
+free th env =
+  maybe (return Nothing)
+        (\par -> do
+          let sibling = children par
+          found <- atomically $ do
+            sbs <- readTVar sibling
+            let (sbs', found) = drop [] th sbs
+            -- !!> "search "++show th ++ " in " ++ show (map threadId sbs)
+            when found $ writeTVar sibling sbs'
+            -- !> ("new list",map threadId sbs')
+            return found
+          if not found
+          then free th par         -- !!> "toparent"
+          else return $ Just env)
+        (parent env)
+   where drop processed th []  = (processed, False)
+         drop processed th (ev:evts)
+           | th == threadId ev = (processed ++ evts, True)
+           -- WARNING: The order of the siblings is not preserved.
+           -- This shouldn't matter, but if it does, look here.
+           | otherwise         = drop (ev:processed) th evts
+
+-- | If the parent is a not a free thread, add a child context to the parent context
+-- atomically.
+hangThread :: EventF -> EventF -> IO ()
+hangThread parent child =
+  when (not $ freeTh parent) $ do
+    let headpths = children parent
+    atomically $ do
+      ths <- readTVar headpths
+      writeTVar headpths $ child:ths
+    -- !!>  "thread added: "++ show (threadId child)
+
+-- | Kill all the child threads associated with the continuation context and reset
+-- the children for the continuation context.
+killChildren :: TVar [EventF] -> IO ()
 killChildren childs  = do
-
 --     forkIO $ do
-        ths <- atomically $ do
-           ths <- readTVar childs
-           writeTVar childs []
-           return ths
+  ths <- atomically $ do
+      ths <- readTVar childs
+      writeTVar childs []
+      return ths
 --        mapM_ killChildren ths       -- recursive not needed, event handlers do it
-
-        mapM_ (killThread . threadId) ths   -- !!> ("KILLEVENT " ++ show (map threadId ths) ++
---                                                        if length ths <20 then ""
---                                                          else error "long list of threads" )
+  mapM_ (killThread . threadId) ths
+  -- !!> ("KILLEVENT " ++ show (map threadId ths) ++
+  -- if length ths <20 then ""
+  --   else error "long list of threads" )
 --     return ()
```

Comment:
Thanks. All improvements are welcome.  

@geraldus I would remove the copyright info since it is weird for open source code. I would change it for "Author"